posts/frangeris/06.stale-cache.md

```diff
@@ -0,0 +1,129 @@
+
+Having understood the concept of stale cache, let's see how we can implement this on a real-world scenario.
+
+## ⚡️ Some testings
+
+Let's test this concept with a simple API that returns a random number, and it will be deployed on `us-east-2` zone on AWS API Gateway.
+
+Making a GET request to the endpoint `/random` will return a random number between 1 and 100, due to the cold-start time of serverless functions, the first request will take a little longer to respond, but the next requests will be faster.
+
+![Response time of first request](./assets/06.image.3.png)
+
+This example shows some of the factors explained at the beginning of the article, subsequent requests can take between `200ms` and `450ms` of response time depending of the lambda invoke response, even with some of tracing being cached like _DNS Lookup, TCP Handshake, SSL Handshake_, this is pretty normal as the origin of the request is from South America and edge locations are not configured.
+
+[image of map travel from ar to east-2]
+
+So, how can we improve this response time? Let's add a cache system to the API.
+
+## Let's introduce bunny.net 🐇
+
+Bunny is a global content delivery platform that help us to delivers cached content faster. In this example, we will use Bunny to cache the responses of the API and see how the cache system works and how the stale cache can be implemented. They have a lot of features and capabilities, but for this example, we will focus on CDN.
+
+With an account already created, we can create a new "pull zone", this is the layer that sits in front of the API and act as intermediary caching everything being requested. Given the name `xxxxxxx.b-cdn.net`, a domain will be created, we need to also add the "Origin URL" that is the api endpoint we want to cache.
+
+While creating the pull zone, they provide the options to check "in what zones" we want to store the cached data, so the content is served from the closest region instead of traveling all the way to the origin, we will select all the zones available so the cache is everywhere 🤯.
+
+![Pull zone for edge locations](./assets/06.image.4.png)
+
+With the pull zone created, is ready to be used as proxy for the API, now lets move back the the endpoint code to make the header customization so bunny know what to do with our origin responses.
+
+We can start making request to the new domain pull zone domain, so any requests made to CDN will be forwarded behind scenes to API but caching the response.
+
+```plaintext
+// instead of
+GET https://api.example.com/random
+
+// use the pull zone domain
+GET https://mynewpullzone.b-cdn.net/random
+```
+
+By default our cache control header is se to `no-cache`, so every time we hit the pull zone url provided it will make a request to the origin, so lets change that behavior by just adding `max-age` directive to the header.
```

Comment:
```suggestion
By default our cache control header is set to `no-cache`, so every time we hit the pull zone URL provided it will make a request to the origin, so lets change that behavior by just adding `max-age` directive to the header.
```